source/grid/manifold_lib.cc

```diff
@@ -361,80 +361,231 @@ get_tangent_vector (const Point<spacedim> &p1,
+  // Do more expensive Newton-style iterations to improve the estimate.
+
+  // Search for duplicate directions and merge them to minimize the cost of
+  // the get_new_point function call below.
+  Table<2,double> merged_weights(weights);
+  boost::container::small_vector<Tensor<1, spacedim>, 100> merged_directions(surrounding_points.size(),Point<spacedim>());
+  boost::container::small_vector<double, 100>              merged_distances(surrounding_points.size(),0.0);
+
+  unsigned int n_unique_directions = 0;
+  for (unsigned int i = 0; i < surrounding_points.size(); ++i)
+    {
+      bool found_duplicate = false;
+
+      // This inner loop if of N^2 complexity, but surrounding_points.size()
+      // is usually at most 8 points large.
+      for (unsigned int j = 0; j < n_unique_directions; ++j)
+        {
+          const double squared_distance = (directions[i] - directions[j]).norm_square();
+          if (!found_duplicate && squared_distance < 1e-28)
+            {
+              found_duplicate = true;
+              for (unsigned int row = 0; row < weights.size(0); ++row)
+                merged_weights[row][j] += weights[row][i];
+            }
+        }
+
+      if (found_duplicate == false)
+        {
+          merged_directions[n_unique_directions] = directions[i];
+          merged_distances[n_unique_directions] = distances[i];
+          for (unsigned int row = 0; row < weights.size(0); ++row)
+            merged_weights[row][n_unique_directions] = weights[row][i];
+
+          ++n_unique_directions;
+        }
+    }
+
+  // Note that we only use the n_unique_directions first entries in the ArrayView
+  const ArrayView<const Tensor<1,spacedim>> array_merged_directions = make_array_view(merged_directions.begin(),
+                                         merged_directions.begin()+n_unique_directions);
+  const ArrayView<const double> array_merged_distances = make_array_view(merged_distances.begin(),
+                                                         merged_distances.begin()+n_unique_directions);
+
+  for (unsigned int row=0; row<weights.size(0); ++row)
+    if (!accurate_point_was_found[row])
+      {
+        const ArrayView<const double> array_merged_weights (merged_weights[row].begin(),n_unique_directions);
+        new_points[row] = get_new_point(array_merged_directions,
+                                        array_merged_distances,
+                                        array_merged_weights,
+                                        new_points[row]);
+      }
+}
+
+
+
 template <int dim, int spacedim>
 Point<spacedim>
 SphericalManifold<dim,spacedim>::
 get_new_point (const ArrayView<const Point<spacedim>> &vertices,
                const ArrayView<const double>          &weights) const
 {
-  const unsigned int n_points = vertices.size();
+  // To avoid duplicating all of the logic in get_new_points, simply call it
+  // for one position.
+  const Table<2,double> table_weights(1,weights.size(),weights.begin());
```

Comment:
It is a bit unfortunate that we need to allocate memory here. Maybe create an internal function that does not take a `Table` as the argument but a plain pointer/`Arrayview`. (We will need to index into those in a somewhat more clumsy way, but @drwells worked hard to avoid memory allocations from these spots...)